File: cli_launcher.py
#!/usr/bin/env python3
import os
import subprocess
import sys


def usage() -> str:
    return (
        "Usage: cli_launcher.py [-p|--port <port>] [-h|--headless] "
        "[-t|--tool <listen|sca>] [-- <sca args>]"
    )


def resolve_script(script_base: str) -> str:
    script_dir = os.path.dirname(os.path.abspath(__file__))
    if os.name == "nt":
        return os.path.join(script_dir, f"{script_base}.bat")
    return os.path.join(script_dir, f"{script_base}.sh")


def parse_args(argv: list[str]) -> tuple[str, str | None, bool, list[str]]:
    tool = None
    port = None
    headless = False
    sca_args: list[str] = []
    index = 0

    while index < len(argv):
        arg = argv[index]
        if arg in ("-t", "--tool"):
            if index + 1 >= len(argv):
                raise ValueError(f"Missing value for {arg}.")
            tool = argv[index + 1]
            index += 2
            if tool == "sca":
                sca_args = argv[index:]
                break
            continue
        if arg in ("-p", "--port"):
            if index + 1 >= len(argv):
                raise ValueError(f"Missing value for {arg}.")
            port = argv[index + 1]
            index += 2
            continue
        if arg in ("-h", "--headless"):
            headless = True
            index += 1
            continue
        if arg == "--help":
            print(usage())
            sys.exit(0)
        if arg == "--":
            sca_args = argv[index + 1 :]
            break
        raise ValueError(f"Unknown option: {arg}")

    if tool is None:
        tool = "listen"

    if tool == "sca" and sca_args[:1] == ["--"]:
        sca_args = sca_args[1:]

    if port and not port.isdigit():
        raise ValueError("Port must be a number.")

    return tool, port, headless, sca_args


def run() -> int:
    try:
        tool, port, headless, sca_args = parse_args(sys.argv[1:])
    except ValueError as exc:
        print(str(exc), file=sys.stderr)
        print(usage(), file=sys.stderr)
        return 2

    print("ADA Scanner CLI Launcher")
    print(f"Selected tool: {tool}")

    if tool == "sca":
        if port or headless:
            print("Warning: --port/--headless ignored for sca.", file=sys.stderr)
        if sca_args:
            print("Static analysis args: " + " ".join(sca_args))
        else:
            print("Static analysis args: (none)")
        script = resolve_script("sca")
        return subprocess.call([script, *sca_args])

    if tool == "listen":
        if port:
            print(f"Port override: {port}")
        else:
            print("Port override: (default)")
        if headless:
            print("Headless: enabled (monitoring console disabled).")
        else:
            print("Headless: disabled")
        args: list[str] = []
        if port:
            args.extend(["--port", port])
        if headless:
            args.append("--headless")
        script = resolve_script("listen")
        return subprocess.call([script, *args])

    print(f"Unknown tool: {tool} (expected listen or sca).", file=sys.stderr)
    return 2


if __name__ == "__main__":
    raise SystemExit(run())

File: CliLauncher.java
import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

public class CliLauncher {
    private static String usage() {
        return "Usage: CliLauncher [-p|--port <port>] [-h|--headless] [-t|--tool <listen|sca>] [-- <sca args>]";
    }

    private static boolean isWindows() {
        return System.getProperty("os.name").toLowerCase().contains("win");
    }

    private static Path scriptDirectory() throws URISyntaxException {
        Path location = Paths.get(CliLauncher.class.getProtectionDomain().getCodeSource().getLocation().toURI());
        if (Files.isRegularFile(location)) {
            return location.getParent();
        }
        return location;
    }

    private static Path resolveScript(String base) throws URISyntaxException {
        String extension = isWindows() ? ".bat" : ".sh";
        return scriptDirectory().resolve(base + extension);
    }

    public static void main(String[] args) throws IOException, URISyntaxException, InterruptedException {
        String tool = null;
        String port = null;
        boolean headless = false;
        List<String> scaArgs = new ArrayList<>();
        boolean stopParsing = false;

        for (int i = 0; i < args.length; i++) {
            String arg = args[i];
            switch (arg) {
                case "-t":
                case "--tool":
                    if (i + 1 >= args.length) {
                        System.err.println("Missing value for " + arg + ".");
                        System.err.println(usage());
                        System.exit(2);
                    }
                    tool = args[++i];
                    if ("sca".equals(tool)) {
                        for (int j = i + 1; j < args.length; j++) {
                            scaArgs.add(args[j]);
                        }
                        stopParsing = true;
                    }
                    break;
                case "-p":
                case "--port":
                    if (i + 1 >= args.length) {
                        System.err.println("Missing value for " + arg + ".");
                        System.err.println(usage());
                        System.exit(2);
                    }
                    port = args[++i];
                    break;
                case "-h":
                case "--headless":
                    headless = true;
                    break;
                case "--":
                    for (int j = i + 1; j < args.length; j++) {
                        scaArgs.add(args[j]);
                    }
                    stopParsing = true;
                    break;
                case "--help":
                    System.out.println(usage());
                    return;
                default:
                    System.err.println("Unknown option: " + arg);
                    System.err.println(usage());
                    System.exit(2);
            }
            if (stopParsing) {
                break;
            }
        }

        if (tool == null) {
            tool = "listen";
        }

        if (port != null && !port.matches("\\d+")) {
            System.err.println("Port must be a number.");
            System.exit(2);
        }

        System.out.println("ADA Scanner CLI Launcher");
        System.out.println("Selected tool: " + tool);

        if ("sca".equals(tool)) {
            if (!scaArgs.isEmpty() && "--".equals(scaArgs.get(0))) {
                scaArgs.remove(0);
            }
            if (port != null || headless) {
                System.err.println("Warning: --port/--headless ignored for sca.");
            }
            if (scaArgs.isEmpty()) {
                System.out.println("Static analysis args: (none)");
            } else {
                System.out.println("Static analysis args: " + String.join(" ", scaArgs));
            }
            Path script = resolveScript("sca");
            List<String> command = new ArrayList<>();
            command.add(script.toString());
            command.addAll(scaArgs);
            Process process = new ProcessBuilder(command).inheritIO().start();
            System.exit(process.waitFor());
        }

        if ("listen".equals(tool)) {
            if (port != null) {
                System.out.println("Port override: " + port);
            } else {
                System.out.println("Port override: (default)");
            }
            if (headless) {
                System.out.println("Headless: enabled (monitoring console disabled).");
            } else {
                System.out.println("Headless: disabled");
            }
            Path script = resolveScript("listen");
            List<String> command = new ArrayList<>();
            command.add(script.toString());
            if (port != null) {
                command.add("--port");
                command.add(port);
            }
            if (headless) {
                command.add("--headless");
            }
            Process process = new ProcessBuilder(command).inheritIO().start();
            System.exit(process.waitFor());
        }

        System.err.println("Unknown tool: " + tool + " (expected listen or sca).");
        System.exit(2);
    }
}
